File: ios/ExpoDynalinksSdkModule.swift
import ExpoModulesCore
import DynalinksSDK

public class ExpoDynalinksSdkModule: Module {
    public func definition() -> ModuleDefinition {
        Name("ExpoDynalinksSdk")

        AsyncFunction("configure") { (config: [String: Any]) throws in
            try await self.configure(config: config)
        }

        AsyncFunction("checkForDeferredDeepLink") { () -> [String: Any] in
            return try await self.checkForDeferredDeepLink()
        }

        AsyncFunction("resolveLink") { (url: String) -> [String: Any] in
            return try await self.resolveLink(url: url)
        }
    }

    private func configure(config: [String: Any]) async throws {
        guard let clientAPIKey = config["clientAPIKey"] as? String else {
            throw DynalinksConfigError.missingAPIKey
        }

        let baseURLString = config["baseURL"] as? String
        let baseURL = baseURLString.flatMap { URL(string: $0) } ?? URL(string: "https://dynalinks.app/api/v1")!
        let logLevelString = config["logLevel"] as? String ?? "error"
        let allowSimulator = config["allowSimulator"] as? Bool ?? false

        let logLevel: DynalinksLogLevel
        switch logLevelString {
        case "none": logLevel = .none
        case "error": logLevel = .error
        case "warning": logLevel = .warning
        case "info": logLevel = .info
        case "debug": logLevel = .debug
        default: logLevel = .error
        }

        do {
            try Dynalinks.configure(
                clientAPIKey: clientAPIKey,
                baseURL: baseURL,
                logLevel: logLevel,
                allowSimulator: allowSimulator
            )
        } catch let error as DynalinksError {
            throw convertError(error)
        }
    }

    private func checkForDeferredDeepLink() async throws -> [String: Any] {
        do {
            let result = try await Dynalinks.checkForDeferredDeepLink()
            return encodeResult(result, isDeferred: true)
        } catch let error as DynalinksError {
            throw convertError(error)
        }
    }

    private func resolveLink(url: String) async throws -> [String: Any] {
        guard let urlObj = URL(string: url) else {
            throw DynalinksResolveError.invalidURL
        }

        do {
            let result = try await Dynalinks.handleUniversalLink(url: urlObj)
            return encodeResult(result, isDeferred: false)
        } catch let error as DynalinksError {
            throw convertError(error)
        }
    }

    private func encodeResult(_ result: DeepLinkResult, isDeferred: Bool) -> [String: Any] {
        var dict: [String: Any] = [
            "matched": result.matched,
            "is_deferred": isDeferred,
        ]

        if let confidence = result.confidence {
            dict["confidence"] = confidence.rawValue
        }

        if let matchScore = result.matchScore {
            dict["match_score"] = matchScore
        }

        if let link = result.link {
            dict["link"] = encodeLinkData(link)
        }

        return dict
    }

    private func encodeLinkData(_ link: DeepLinkResult.LinkData) -> [String: Any?] {
        return [
            "id": link.id,
            "name": link.name,
            "path": link.path,
            "shortened_path": link.shortenedPath,
            "url": link.url?.absoluteString,
            "full_url": link.fullURL?.absoluteString,
            "deep_link_value": link.deepLinkValue,
            "ios_fallback_url": link.iosFallbackURL?.absoluteString,
            "android_fallback_url": link.androidFallbackURL?.absoluteString,
            "enable_forced_redirect": link.enableForcedRedirect,
            "social_title": link.socialTitle,
            "social_description": link.socialDescription,
            "social_image_url": link.socialImageURL?.absoluteString,
            "clicks": link.clicks,
        ]
    }

    private func convertError(_ error: DynalinksError) -> Exception {
        let code: String
        let message: String

        switch error {
        case .notConfigured:
            code = "NOT_CONFIGURED"
            message = error.errorDescription ?? "SDK not configured"
        case .invalidAPIKey(let msg):
            code = "INVALID_API_KEY"
            message = msg
        case .simulator:
            code = "SIMULATOR"
            message = error.errorDescription ?? "Not available on simulator"
        case .networkError:
            code = "NETWORK_ERROR"
            message = error.errorDescription ?? "Network error"
        case .invalidResponse:
            code = "INVALID_RESPONSE"
            message = error.errorDescription ?? "Invalid response"
        case .serverError(let statusCode, let msg):
            code = "SERVER_ERROR"
            message = msg ?? "Server error: \(statusCode)"
        case .noMatch:
            code = "NO_MATCH"
            message = error.errorDescription ?? "No match"
        }

        return Exception(name: code, description: message)
    }
}

enum DynalinksConfigError: Error {
    case missingAPIKey
}

enum DynalinksResolveError: Error {
    case invalidURL
}

File: android/src/main/java/expo/modules/dynalinkssdk/ExpoDynalinksSdkModule.kt
package expo.modules.dynalinkssdk

import android.net.Uri
import expo.modules.kotlin.modules.Module
import expo.modules.kotlin.modules.ModuleDefinition
import expo.modules.kotlin.exception.CodedException
import com.dynalinks.sdk.Dynalinks
import com.dynalinks.sdk.DeepLinkResult
import com.dynalinks.sdk.DynalinksError
import com.dynalinks.sdk.DynalinksLogLevel
import com.dynalinks.sdk.LinkData

class ExpoDynalinksSdkModule : Module() {
  override fun definition() = ModuleDefinition {
    Name("ExpoDynalinksSdk")

    AsyncFunction("configure") { config: Map<String, Any?> ->
      configure(config)
    }

    AsyncFunction("checkForDeferredDeepLink") {
      checkForDeferredDeepLink()
    }

    AsyncFunction("resolveLink") { url: String ->
      resolveLink(url)
    }
  }

  private suspend fun configure(config: Map<String, Any?>) {
    val context = appContext.reactContext ?: throw CodedException("Context not available")

    val clientAPIKey = (config["clientAPIKey"] as? String)?.takeIf { it.isNotBlank() }
      ?: throw CodedException("INVALID_API_KEY", "Missing or empty clientAPIKey", null)

    val logLevelString = config["logLevel"] as? String ?: "error"
    val allowSimulator = config["allowSimulator"] as? Boolean ?: false

    val logLevel = when (logLevelString) {
      "none" -> DynalinksLogLevel.NONE
      "error" -> DynalinksLogLevel.ERROR
      "warning" -> DynalinksLogLevel.WARNING
      "info" -> DynalinksLogLevel.INFO
      "debug" -> DynalinksLogLevel.DEBUG
      else -> DynalinksLogLevel.ERROR
    }

    try {
      // Only pass baseURL if provided - let native SDK use its default otherwise
      val baseURLString = (config["baseURL"] as? String)?.takeIf { it.isNotBlank() }

      if (baseURLString != null) {
        // Validate baseURL format if provided
        val baseUri = Uri.parse(baseURLString)
        if (baseUri.scheme.isNullOrEmpty() || baseUri.host.isNullOrEmpty()) {
          throw CodedException("INVALID_CONFIG", "Invalid baseURL format", null)
        }

        Dynalinks.configure(
          context = context,
          clientAPIKey = clientAPIKey,
          baseURL = baseURLString,
          logLevel = logLevel,
          allowEmulator = allowSimulator
        )
      } else {
        // Use native SDK's default baseURL
        Dynalinks.configure(
          context = context,
          clientAPIKey = clientAPIKey,
          logLevel = logLevel,
          allowEmulator = allowSimulator
        )
      }
    } catch (e: DynalinksError) {
      throw convertError(e)
    }
  }

  private suspend fun checkForDeferredDeepLink(): Map<String, Any?> {
    return try {
      val result = Dynalinks.checkForDeferredDeepLink()
      encodeResult(result, isDeferred = true)
    } catch (e: DynalinksError) {
      throw convertError(e)
    }
  }

  private suspend fun resolveLink(url: String): Map<String, Any?> {
    return try {
      val uri = Uri.parse(url)

      // Validate the parsed URI
      if (uri.scheme.isNullOrEmpty() || uri.host.isNullOrEmpty()) {
        throw CodedException("INVALID_URL", "URL must have a valid scheme and host", null)
      }

      val result = Dynalinks.handleAppLink(uri)
      encodeResult(result, isDeferred = false)
    } catch (e: CodedException) {
      throw e
    } catch (e: DynalinksError) {
      throw convertError(e)
    } catch (e: Exception) {
      throw CodedException("UNKNOWN", "Unexpected error: ${e.message}", e)
    }
  }

  private fun encodeResult(result: DeepLinkResult, isDeferred: Boolean): Map<String, Any?> {
    val map = mutableMapOf<String, Any?>(
      "matched" to result.matched,
      "is_deferred" to isDeferred
    )

    result.confidence?.let {
      map["confidence"] = it.name.lowercase()
    }

    result.matchScore?.let {
      map["match_score"] = it
    }

    result.link?.let { link ->
      map["link"] = encodeLinkData(link)
    }

    return map
  }

  private fun encodeLinkData(linkData: LinkData): Map<String, Any?> {
    return mapOf(
      "id" to linkData.id,
      "name" to linkData.name,
      "path" to linkData.path,
      "shortened_path" to linkData.shortenedPath,
      "url" to linkData.url,
      "full_url" to linkData.fullUrl,
      "deep_link_value" to linkData.deepLinkValue,
      "android_fallback_url" to linkData.androidFallbackUrl,
      "ios_fallback_url" to linkData.iosFallbackUrl,
      "enable_forced_redirect" to linkData.enableForcedRedirect,
      "social_title" to linkData.socialTitle,
      "social_description" to linkData.socialDescription,
      "social_image_url" to linkData.socialImageUrl,
      "clicks" to linkData.clicks
    ).filterValues { it != null }
  }

  private fun convertError(error: DynalinksError): CodedException {
    val code = when (error) {
      is DynalinksError.NotConfigured -> "NOT_CONFIGURED"
      is DynalinksError.InvalidAPIKey -> "INVALID_API_KEY"
      is DynalinksError.Emulator -> "SIMULATOR"
      is DynalinksError.InvalidIntent -> "INVALID_URL"
      is DynalinksError.NetworkError -> "NETWORK_ERROR"
      is DynalinksError.InvalidResponse -> "INVALID_RESPONSE"
      is DynalinksError.ServerError -> "SERVER_ERROR"
      is DynalinksError.NoMatch -> "NO_MATCH"
      is DynalinksError.InstallReferrerUnavailable -> "INSTALL_REFERRER_UNAVAILABLE"
      is DynalinksError.InstallReferrerTimeout -> "INSTALL_REFERRER_TIMEOUT"
    }
    return CodedException(code, error.message ?: "Unknown error", null)
  }
}
